{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import csv\n",
    "from sympy import FiniteSet\n",
    "\n",
    "#import pysam\n",
    "import os\n",
    "import math\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "from pathlib import Path\n",
    "from random import seed\n",
    "from random import randint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#funciones\n",
    "\n",
    "def powerset(fullset):\n",
    "  listsub = list(fullset)\n",
    "  subsets = []\n",
    "  for i in range(2**len(listsub)):\n",
    "    subset = []\n",
    "    for k in range(len(listsub)):            \n",
    "      if i & 1<<k:\n",
    "        subset.append(listsub[k])\n",
    "    subsets.append(subset)        \n",
    "  return subsets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vamos a mapear los reads verdaderos de maiz a los genomas de la base de datos interna de referencia de _Clavibacter_ (BDIRC) \n",
    "0. Crear link simbolicos\n",
    "1. Indexar los genomas de BDIRC con bowtie\n",
    "2. Mapear con bowtie a esos genomas\n",
    "3. Si queremos visualizar los reads podemos usar tablet, y necesitamos samtools para indexar el genoma y ver los reads alineados.\n",
    "https://www.metagenomics.wiki/tools/bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A continuaciÃ³n creamos indices con para cada uno de los genomas refencia, para ello utilizamos el comando `bowtie2-build`. Los respectivos archivos de salida los almacenamos en la carpeta `index-bowtie`. Adicionalmente, creamos el archivo `list-genomes.txt` que contiene los nombre de cada archivo de salida del `bowtie2-build`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1101.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1106.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1207.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_PF008.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_ATCC_10253.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_1195.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_2404.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_6488.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_LMG_3663.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-3.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_419B.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_44.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_61-1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_7580.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_A6096.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CFBP_7577.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CIBA.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_395.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_397.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_HF4.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_NCPPB_2581.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_SL1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_ATCC33113.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-Cs3N.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-CsR14.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_ATCC_33566.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_DOAB_609.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1101.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1106.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1207.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_PF008.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_ATCC_10253.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_1195.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_2404.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_6488.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_LMG_3663.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-3.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_419B.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_44.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_61-1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_7580.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_A6096.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CFBP_7577.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CIBA.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_395.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_397.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_HF4.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_NCPPB_2581.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_SL1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_ATCC33113.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-Cs3N.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-CsR14.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_ATCC_33566.fna genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_DOAB_609.fna genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.4.bt2\" file written during aborted indexing attempt.\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "ls genomas-clavi/clavi-genomes/*fna | while read line; \n",
    "do \n",
    "    id=$(echo $line| rev| cut -d'_' -f1,2,3 |rev |cut -d'.' -f1);\n",
    "    bowtie2-build $line genomas-clavi/clavi-genomes/index-bowtie/$id;\n",
    "    echo \"$id\" >>list-genomes.txt\n",
    "done\n",
    "\n",
    "#Por ahora solo corre en bash\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lo siguiente es mapear los reads a los genomas de referencia, utilizamos `bowtie2` y obtenemos arhivos de salidas *.sam* que guardaremos en la carpeta `output-bowtie` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#reference=subsp_capsici_1101  ; usar para extraer las lineas de archivo sed '1q;d' index-genomes.txt\n",
    "for infile in ex-reads-clavi/fastq/fastq/capsicum/*-clav-1.fq\n",
    " do \n",
    "base=$(basename ${infile} -clav-1.fq)\n",
    "dir=$(dirname ${infile} )\n",
    "bowtie2 -x genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101 -1 ${dir}/${base}-clav-1.fq -2 ${dir}/${base}-clav-2.fq  --no-unal -p 12 -S genomas-clavi/clavi-genomes/out-bowtie/${base}.sam\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "bash 01-mapear-reads-a-referencia.sh insidiosus_CFBP_2404"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora corremos el scrip `01-mapear-reads-a-referencia.sh` por cada una de la lista de genomas que se encuentra en `index-genomes.txtÂ´ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat index-genomes.txt | while read line;\n",
    "do\n",
    "    bash 01-mapear-reads-a-referencia.sh $line\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prueba 1: Utilizar solo 3 genomas\n",
    "Ahora procedor hacer lo anterior son con 3 genomas, 1 de Clavibacter Michiganensis Michiganensis (cmm), y 2 de Michagenesis capsisi. \n",
    "Primero cremos los index\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1101.fna-cortado.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1106.fna-cortado.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_michi_contigs.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: /home/shaday/anaconda3/etc/profile.d/conda.sh: No such file or directory\n",
      "\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.\n",
      "To initialize your shell, run\n",
      "\n",
      "    $ conda init <SHELL_NAME>\n",
      "\n",
      "Currently supported shells are:\n",
      "  - bash\n",
      "  - fish\n",
      "  - tcsh\n",
      "  - xonsh\n",
      "  - zsh\n",
      "  - powershell\n",
      "\n",
      "See 'conda init --help' for more information and options.\n",
      "\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\n",
      "\n",
      "\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1101.fna-cortado.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101 \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1106.fna-cortado.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106 \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_michi_contigs.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.4.bt2\" file written during aborted indexing attempt.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "ls genomas-clavi/clavi-genomes/genomas-prueba1/*fna | while read line; \n",
    "do \n",
    "    id=$(echo $line| rev| cut -d'_' -f1,2,3 |rev |cut -d'.' -f1);\n",
    "    bowtie2-build $line genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/$id;\n",
    "    echo \"$id\" >>list-genomasrefencia.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora, hacemos el mapeo de los read simulados a cada genoma de refencia previamente indexado\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.\n",
      "To initialize your shell, run\n",
      "\n",
      "    $ conda init <SHELL_NAME>\n",
      "\n",
      "Currently supported shells are:\n",
      "  - bash\n",
      "  - fish\n",
      "  - tcsh\n",
      "  - xonsh\n",
      "  - zsh\n",
      "  - powershell\n",
      "\n",
      "See 'conda init --help' for more information and options.\n",
      "\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\n",
      "\n",
      "\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "(ERR): Could not open output file 'output-tda/prueba1Kread/out-bowtie/Simulados-subsp_capsici_1101.sam' for writing.\n",
      "Exiting now ...\n",
      "stat: Bad file descriptor\n",
      "Warning: Could not open read file \"reads-sim1K.fastq.fastq\" for reading; skipping...\n",
      "Error: No input read files were valid\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "(ERR): Could not open output file 'output-tda/prueba1Kread/out-bowtie/Simulados-subsp_capsici_1106.sam' for writing.\n",
      "Exiting now ...\n",
      "stat: Bad file descriptor\n",
      "Warning: Could not open read file \"reads-sim1K.fastq.fastq\" for reading; skipping...\n",
      "Error: No input read files were valid\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 423.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value in exists at /opt/anaconda3/bin/bowtie2 line 81.\n",
      "Use of uninitialized value $bt2_args[4] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "Use of uninitialized value $bt2_args[5] in join or string at /opt/anaconda3/bin/bowtie2 line 459.\n",
      "(ERR): Could not open output file 'output-tda/prueba1Kread/out-bowtie/Simulados-subsp_michi_contigs.sam' for writing.\n",
      "Exiting now ...\n",
      "stat: Bad file descriptor\n",
      "Warning: Could not open read file \"reads-sim1K.fastq.fastq\" for reading; skipping...\n",
      "Error: No input read files were valid\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'#source ~/anaconda3/etc/profile.d/conda.sh\\nconda activate metagenomics\\n\\ncat list-genomasrefencia.txt | while read line;\\ndo\\nbowtie2 -x output-tda/prueba-3genomas/index-bowtie/$line -U reads-sim1K.fastq.fastq --no-unal -p 12 -S output-tda/prueba1Kread/out-bowtie/Simulados-${line}.sam\\ndone\\n'' returned non-zero exit status 13.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-82e83d8b5bac>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bash'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'#source ~/anaconda3/etc/profile.d/conda.sh\\nconda activate metagenomics\\n\\ncat list-genomasrefencia.txt | while read line;\\ndo\\nbowtie2 -x output-tda/prueba-3genomas/index-bowtie/$line -U reads-sim1K.fastq.fastq --no-unal -p 12 -S output-tda/prueba1Kread/out-bowtie/Simulados-${line}.sam\\ndone\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2397\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2398\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2399\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2400\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2401\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/decorator.py\u001b[0m in \u001b[0;36mfun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    229\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mkwsyntax\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    230\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 231\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mcaller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextras\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    232\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    233\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'#source ~/anaconda3/etc/profile.d/conda.sh\\nconda activate metagenomics\\n\\ncat list-genomasrefencia.txt | while read line;\\ndo\\nbowtie2 -x output-tda/prueba-3genomas/index-bowtie/$line -U reads-sim1K.fastq.fastq --no-unal -p 12 -S output-tda/prueba1Kread/out-bowtie/Simulados-${line}.sam\\ndone\\n'' returned non-zero exit status 13."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "\n",
    "cat list-genomasrefencia.txt | while read line;\n",
    "do\n",
    "bowtie2 -x output-tda/prueba-3genomas/index-bowtie/$line -U reads-sim1K.fastq.fastq --no-unal -p 12 -S output-tda/prueba1Kread/out-bowtie/Simulados-${line}.sam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Crear archivos .bam para visualizar y extraer informaciÃ³n del match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for infile in output-tda/prueba1Kread/out-bowtie/*.sam\n",
    "do \n",
    "\tbase=$(basename ${infile} .sam)\n",
    "\tsamtools view ${infile} -o output-tda/prueba1Kread/bam/${base}.bam\n",
    "    samtools sort ${infile} -o output-tda/prueba1Kread/sorted/${base}_sorted.sam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Crea un archivo *.txt con la salida mostrando el identificador y el read que identifico en el genoma referencia."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for infile in output-tda/prueba1Kread/bam/*.bam\n",
    "do \n",
    "\tbase=$(basename ${infile} .bam)\n",
    "\tsamtools sort ${infile} -o output-tda/prueba1Kread/sorted/${base}_sorted.bam\n",
    "\tsamtools view ${infile} > output-tda/prueba1Kread/bam/${base}.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obtener las columnas que nos interesan de los archivos .sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ~/GIT/clavibacter/scripts-tda/\n",
    "ls output-tda/prueba1Kread/bam/*.txt | while read line\n",
    "do\n",
    "    base=$(basename ${line} .txt)\n",
    "    cut -f 1-6 $line > output-tda/prueba1Kread/tablas/${base}_extrac.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Importar tablas de mapeos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Qname</th>\n",
       "      <th>Flag</th>\n",
       "      <th>Rname</th>\n",
       "      <th>Pos</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>Organims</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_009478.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>578115</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_009480.1_1</td>\n",
       "      <td>16</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>133457</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_009478.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>16945</td>\n",
       "      <td>38</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_009479.1_1</td>\n",
       "      <td>16</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>42991</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_CP048047.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>27280</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4397</th>\n",
       "      <td>NC_009478.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>12053</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4398</th>\n",
       "      <td>NC_009479.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>1952</td>\n",
       "      <td>0</td>\n",
       "      <td>122M1D28M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4399</th>\n",
       "      <td>NZ_CP048049.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2365649</td>\n",
       "      <td>3</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4400</th>\n",
       "      <td>NC_009480.1_994</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>81133</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4401</th>\n",
       "      <td>NC_009478.1_994</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>1549686</td>\n",
       "      <td>8</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13942 rows Ã 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Qname  Flag          Rname      Pos  MapQ      CIGAR  \\\n",
       "0         NC_009478.1_0     0  NZ_CP048047.1   578115    42       150M   \n",
       "1         NC_009480.1_1    16  NZ_CP048048.1   133457    42       150M   \n",
       "2         NC_009478.1_1     0  NZ_CP048048.1    16945    38       150M   \n",
       "3         NC_009479.1_1    16  NZ_CP048048.1    42991    42       150M   \n",
       "4       NZ_CP048047.1_1     0  NZ_CP048047.1    27280    42       150M   \n",
       "...                 ...   ...            ...      ...   ...        ...   \n",
       "4397    NC_009478.1_993     0    NC_009479.1    12053    42       150M   \n",
       "4398    NC_009479.1_993     0    NC_009479.1     1952     0  122M1D28M   \n",
       "4399  NZ_CP048049.1_993     0    NC_009480.1  2365649     3       150M   \n",
       "4400    NC_009480.1_994     0    NC_009480.1    81133    42       150M   \n",
       "4401    NC_009478.1_994     0    NC_009480.1  1549686     8       150M   \n",
       "\n",
       "           Organims  \n",
       "0      capsici_1106  \n",
       "1      capsici_1106  \n",
       "2      capsici_1106  \n",
       "3      capsici_1106  \n",
       "4      capsici_1106  \n",
       "...             ...  \n",
       "4397  michi_contigs  \n",
       "4398  michi_contigs  \n",
       "4399  michi_contigs  \n",
       "4400  michi_contigs  \n",
       "4401  michi_contigs  \n",
       "\n",
       "[13942 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ruta_tablas = '/home/shaday/GIT/clavibacter/scripts-tda/output-tda/prueba1Kread/tablas/'\n",
    "contenido = os.listdir(ruta_tablas)\n",
    "genomes = []\n",
    "df=[]\n",
    "for i in contenido:\n",
    "    dftt = pd.read_csv(ruta_tablas + i, sep='\\t', header=None) #importar las tablas para cada genoma\n",
    "    dftt['Organims']=i[16:-11] #agregar el nombre del organismo\n",
    "    df.append(dftt) #agregar al dataframe\n",
    "    genomes.append(i[16:-11])\n",
    "    \n",
    "df=pd.concat(df)\n",
    "df.columns= ['Qname','Flag', 'Rname', 'Pos','MapQ','CIGAR','Organims'] #cambiar nombre a las columnas del dataframe\n",
    "df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('MapeoReads-organismos.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creamos un diccionario para los organimos, y remplazamos en el dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Qname</th>\n",
       "      <th>Flag</th>\n",
       "      <th>Rname</th>\n",
       "      <th>Pos</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>Organims</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_009478.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>578115</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_009480.1_1</td>\n",
       "      <td>16</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>133457</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_009478.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>16945</td>\n",
       "      <td>38</td>\n",
       "      <td>150M</td>\n",
       "      <td>X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_009479.1_1</td>\n",
       "      <td>16</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>42991</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_CP048047.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>27280</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4397</th>\n",
       "      <td>NC_009478.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>12053</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4398</th>\n",
       "      <td>NC_009479.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>1952</td>\n",
       "      <td>0</td>\n",
       "      <td>122M1D28M</td>\n",
       "      <td>X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4399</th>\n",
       "      <td>NZ_CP048049.1_993</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2365649</td>\n",
       "      <td>3</td>\n",
       "      <td>150M</td>\n",
       "      <td>X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4400</th>\n",
       "      <td>NC_009480.1_994</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>81133</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>X3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4401</th>\n",
       "      <td>NC_009478.1_994</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>1549686</td>\n",
       "      <td>8</td>\n",
       "      <td>150M</td>\n",
       "      <td>X3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13942 rows Ã 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Qname  Flag          Rname      Pos  MapQ      CIGAR  \\\n",
       "0         NC_009478.1_0     0  NZ_CP048047.1   578115    42       150M   \n",
       "1         NC_009480.1_1    16  NZ_CP048048.1   133457    42       150M   \n",
       "2         NC_009478.1_1     0  NZ_CP048048.1    16945    38       150M   \n",
       "3         NC_009479.1_1    16  NZ_CP048048.1    42991    42       150M   \n",
       "4       NZ_CP048047.1_1     0  NZ_CP048047.1    27280    42       150M   \n",
       "...                 ...   ...            ...      ...   ...        ...   \n",
       "4397    NC_009478.1_993     0    NC_009479.1    12053    42       150M   \n",
       "4398    NC_009479.1_993     0    NC_009479.1     1952     0  122M1D28M   \n",
       "4399  NZ_CP048049.1_993     0    NC_009480.1  2365649     3       150M   \n",
       "4400    NC_009480.1_994     0    NC_009480.1    81133    42       150M   \n",
       "4401    NC_009478.1_994     0    NC_009480.1  1549686     8       150M   \n",
       "\n",
       "     Organims  \n",
       "0          X1  \n",
       "1          X1  \n",
       "2          X1  \n",
       "3          X1  \n",
       "4          X1  \n",
       "...       ...  \n",
       "4397       X3  \n",
       "4398       X3  \n",
       "4399       X3  \n",
       "4400       X3  \n",
       "4401       X3  \n",
       "\n",
       "[13942 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X=df['Organims'].unique().tolist()\n",
    "len(X)\n",
    "dic_Organisms={}\n",
    "for k in range(len(X)):\n",
    "    dic_Organisms[X[k]]=\"X{0}\".format(k+1)\n",
    "dic_Organisms\n",
    "df2=df.replace({\"Organims\": dic_Organisms}) \n",
    "\n",
    "X2=df2['Organims'].unique().tolist() #lista de genomas \n",
    "A=powerset(X2)      #conjunto potencia\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vamos a crear un diccionario que contenga los reads y los organismos a los cuales son mapeados"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic={}\n",
    "l=0\n",
    "Reads=df2['Qname'].unique().tolist()\n",
    "for k in Reads:\n",
    "    valores=list(df2.loc[df2.Qname==k].iloc[:,6])\n",
    "    dic[k]=valores\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La funciÃ³n Â´SyÂ´ calcula los reads que son mapeados solamente a $Y\\in2^X$.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sy(dic,X):\n",
    "    dict_items=dic.items()\n",
    "    i=0\n",
    "    for k in X:\n",
    "        myValue=k\n",
    "        globals()[\"Y{}\".format(i)] =[key for key,value in dict_items if value==myValue]\n",
    "        mykey=[key for key,value in dict_items if value==myValue]\n",
    "        i+=1\n",
    "    return Y0,Y1,Y2,Y3,Y4,Y5,Y6,Y7\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculamos SY para cada Y en 2**X\n",
    "SY=[]\n",
    "SY= Sy(dic,A)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Exportamos Sy a txt\n",
    "for k in range(len(SY)):\n",
    "    with open('output-tda/prueba1Kread/SyReads/S_y{0}.txt'.format(k), 'w') as f:\n",
    "        for line in SY[k]:\n",
    "            f.write(f\"{line}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for infile in output-tda/prueba1Kread/sorted/*.sam\n",
    "do \n",
    "\tbase=$(basename ${infile} .sam)\n",
    "    head -4 ${infile} > output-tda/prueba1Kread/sam_extract/${base}-S_y2.bam  #get the head the .sam file's\n",
    "\tsamtools view ${infile} | grep -f output-tda/prueba1Kread/SyReads/S_y2.txt >> output-tda/prueba1Kread/sam_extract/${base}-S_y2.bam # append the read selected \n",
    "    done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/shaday/GIT/clavibacter/scripts-tda\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "output-tda/prueba1Kread/SyReads/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
