{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import csv\n",
    "from sympy import FiniteSet\n",
    "\n",
    "import pysam\n",
    "import os\n",
    "import math\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "from pathlib import Path\n",
    "from random import seed\n",
    "from random import randint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path="
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vamos a mapear los reads verdaderos de maiz a los genomas de la base de datos interna de referencia de _Clavibacter_ (BDIRC) \n",
    "0. Crear link simbolicos\n",
    "1. Indexar los genomas de BDIRC con bowtie\n",
    "2. Mapear con bowtie a esos genomas\n",
    "3. Si queremos visualizar los reads podemos usar tablet, y necesitamos samtools para indexar el genoma y ver los reads alineados.\n",
    "https://www.metagenomics.wiki/tools/bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A continuaci√≥n creamos indices con para cada uno de los genomas refencia, para ello utilizamos el comando `bowtie2-build`. Los respectivos archivos de salida los almacenamos en la carpeta `index-bowtie`. Adicionalmente, creamos el archivo `list-genomes.txt` que contiene los nombre de cada archivo de salida del `bowtie2-build`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1101.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1106.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1207.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_PF008.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_ATCC_10253.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_1195.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_2404.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_6488.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_LMG_3663.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-3.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_419B.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_44.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_61-1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_7580.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_A6096.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CFBP_7577.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CIBA.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_395.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_397.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_HF4.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_NCPPB_2581.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_SL1.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_ATCC33113.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-Cs3N.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-CsR14.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_ATCC_33566.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_DOAB_609.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1101.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1106.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1106.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_1207.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1207.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_capsici_PF008.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_PF008.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_ATCC_10253.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_ATCC_10253.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_1195.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_1195.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_2404.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_CFBP_6488.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_6488.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_LMG_3663.fna genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_LMG_3663.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_insidiosus_R1-3.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_insidiosus_R1-3.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_419B.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_419B.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_44.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_44.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_61-1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_61-1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_7580.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_7580.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_A6096.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_A6096.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CFBP_7577.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_CFBP_7577.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_CIBA.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_CIBA.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_395.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_395.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_DOAB_397.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_DOAB_397.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_HF4.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_HF4.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_NCPPB_2581.fna genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/nebraskensis_NCPPB_2581.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_nebraskensis_SL1.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_nebraskensis_SL1.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_ATCC33113.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_ATCC33113.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-Cs3N.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-Cs3N.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_sepedonicus_CFIA-CsR14.fna genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/subsp_sepedonicus_CFIA-CsR14.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_ATCC_33566.fna genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_ATCC_33566.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/Clavibacter_michiganensis_subsp_tessellarius_DOAB_609.fna genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609 \n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/index-bowtie/tessellarius_DOAB_609.4.bt2\" file written during aborted indexing attempt.\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "ls genomas-clavi/clavi-genomes/*fna | while read line; \n",
    "do \n",
    "    id=$(echo $line| rev| cut -d'_' -f1,2,3 |rev |cut -d'.' -f1);\n",
    "    bowtie2-build $line genomas-clavi/clavi-genomes/index-bowtie/$id;\n",
    "    echo \"$id\" >>list-genomes.txt\n",
    "done\n",
    "\n",
    "#Por ahora solo corre en bash\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lo siguiente es mapear los reads a los genomas de referencia, utilizamos `bowtie2` y obtenemos arhivos de salidas *.sam* que guardaremos en la carpeta `output-bowtie` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#reference=subsp_capsici_1101  ; usar para extraer las lineas de archivo sed '1q;d' index-genomes.txt\n",
    "for infile in ex-reads-clavi/fastq/fastq/capsicum/*-clav-1.fq\n",
    " do \n",
    "base=$(basename ${infile} -clav-1.fq)\n",
    "dir=$(dirname ${infile} )\n",
    "bowtie2 -x genomas-clavi/clavi-genomes/index-bowtie/subsp_capsici_1101 -1 ${dir}/${base}-clav-1.fq -2 ${dir}/${base}-clav-2.fq  --no-unal -p 12 -S genomas-clavi/clavi-genomes/out-bowtie/${base}.sam\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n",
      "(ERR): \"genomas-clavi/clavi-genomes/index-bowtie/insidiosus_CFBP_2404\" does not exist or is not a Bowtie 2 index\n",
      "Exiting now ...\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "bash 01-mapear-reads-a-referencia.sh insidiosus_CFBP_2404"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora corremos el scrip `01-mapear-reads-a-referencia.sh` por cada una de la lista de genomas que se encuentra en `index-genomes.txt¬¥ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat index-genomes.txt | while read line;\n",
    "do\n",
    "    bash 01-mapear-reads-a-referencia.sh $line\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prueba 1: Utilizar solo 3 genomas\n",
    "Ahora procedor hacer lo anterior son con 3 genomas, 1 de Clavibacter Michiganensis Michiganensis (cmm), y 2 de Michagenesis capsisi. \n",
    "Primero cremos los index\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1101.fna-cortado.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1106.fna-cortado.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Settings:\n",
      "  Output files: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_michi_contigs.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Total time for call to driver() for forward index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1101.fna-cortado.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101 \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1101.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_capsici_1106.fna-cortado.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106 \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_capsici_1106.4.bt2\" file written during aborted indexing attempt.\n",
      "Building a SMALL index\n",
      "Could not open index file for writing: \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.3.bt2\"\n",
      "Please make sure the directory exists and that permissions allow writing by\n",
      "Bowtie.\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\n",
      "Command: /opt/anaconda3/envs/metagenomics/bin/bowtie2-build-s --wrapper basic-0 genomas-clavi/clavi-genomes/genomas-prueba1/Clavibacter_michiganensis_subsp_michi_contigs.fna genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs \n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.3.bt2\" file written during aborted indexing attempt.\n",
      "Deleting \"genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/subsp_michi_contigs.4.bt2\" file written during aborted indexing attempt.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "ls genomas-clavi/clavi-genomes/genomas-prueba1/*fna | while read line; \n",
    "do \n",
    "    id=$(echo $line| rev| cut -d'_' -f1,2,3 |rev |cut -d'.' -f1);\n",
    "    bowtie2-build $line genomas-clavi/clavi-genomes/genomas-prueba1/index-bowtie/$id;\n",
    "    echo \"$id\" >>list-genomasrefencia.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora, hacemos el mapeo de los read simulados a cada genoma de refencia previamente indexado\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    232352 (33.19%) aligned 0 times\n",
      "    460838 (65.83%) aligned exactly 1 time\n",
      "    6810 (0.97%) aligned >1 times\n",
      "66.81% overall alignment rate\n",
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    231193 (33.03%) aligned 0 times\n",
      "    462956 (66.14%) aligned exactly 1 time\n",
      "    5851 (0.84%) aligned >1 times\n",
      "66.97% overall alignment rate\n",
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    254665 (36.38%) aligned 0 times\n",
      "    421192 (60.17%) aligned exactly 1 time\n",
      "    24143 (3.45%) aligned >1 times\n",
      "63.62% overall alignment rate\n",
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    232352 (33.19%) aligned 0 times\n",
      "    460838 (65.83%) aligned exactly 1 time\n",
      "    6810 (0.97%) aligned >1 times\n",
      "66.81% overall alignment rate\n",
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    231193 (33.03%) aligned 0 times\n",
      "    462956 (66.14%) aligned exactly 1 time\n",
      "    5851 (0.84%) aligned >1 times\n",
      "66.97% overall alignment rate\n",
      "700000 reads; of these:\n",
      "  700000 (100.00%) were unpaired; of these:\n",
      "    254665 (36.38%) aligned 0 times\n",
      "    421192 (60.17%) aligned exactly 1 time\n",
      "    24143 (3.45%) aligned >1 times\n",
      "63.62% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source ~/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate metagenomics\n",
    "\n",
    "cat list-genomasrefencia.txt | while read line;\n",
    "do\n",
    "bowtie2 -x output-tda/prueba-3genomas/index-bowtie/$line -U reads-simulados100M.fastq --no-unal -p 12 -S output-tda/prueba-3genomas/out-bowtie/Simulados-${line}.sam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Crear archivos .bam para visualizar y extraer informaci√≥n del match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for infile in output-tda/prueba-3genomas/out-bowtie/*.sam\n",
    "do \n",
    "\tbase=$(basename ${infile} .sam)\n",
    "\tsamtools view -b -S ${infile} -o output-tda/prueba-3genomas/bam/${base}.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Crea un archivo *.txt con la salida mostrando el identificador y el read que identifico en el genoma referencia."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for infile in output-tda/prueba-3genomas/bam/*.bam\n",
    "do \n",
    "\tbase=$(basename ${infile} .bam)\n",
    "\tsamtools sort ${infile} -o output-tda/prueba-3genomas/sorted/${base}_sorted.bam\n",
    "\tsamtools view ${infile} > output-tda/prueba-3genomas/bam/${base}.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obtener las columnas que nos interesan de los archivos .sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ~/GIT/clavibacter/scripts-tda/\n",
    "ls output-tda/prueba-3genomas/bam/*.txt | while read line\n",
    "do\n",
    "    base=$(basename ${line} .txt)\n",
    "    cut -f 1-6 $line > output-tda/prueba-3genomas/tablas/${base}_extrac.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Importar tablas de mapeos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Qname</th>\n",
       "      <th>Flag</th>\n",
       "      <th>Rname</th>\n",
       "      <th>Pos</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>Organims</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048049.1_0</td>\n",
       "      <td>16</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>111591</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC_009478.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>2691532</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC_009479.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>81328</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NZ_CP048048.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048047.1</td>\n",
       "      <td>407360</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NZ_CP048050.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048048.1</td>\n",
       "      <td>44206</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>capsici_1106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445330</td>\n",
       "      <td>NZ_CP048049.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>243049</td>\n",
       "      <td>0</td>\n",
       "      <td>42M1I3M1D104M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445331</td>\n",
       "      <td>NZ_CP048050.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2942610</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445332</td>\n",
       "      <td>NC_009480.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>69488</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445333</td>\n",
       "      <td>NC_009478.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>1808419</td>\n",
       "      <td>0</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445334</td>\n",
       "      <td>NC_009479.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2741813</td>\n",
       "      <td>24</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1381790 rows √ó 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Qname  Flag          Rname      Pos  MapQ  \\\n",
       "0           NZ_CP048049.1_0    16  NZ_CP048048.1   111591    42   \n",
       "1             NC_009478.1_0     0  NZ_CP048047.1  2691532    42   \n",
       "2             NC_009479.1_0     0  NZ_CP048048.1    81328    42   \n",
       "3           NZ_CP048048.1_0     0  NZ_CP048047.1   407360    42   \n",
       "4           NZ_CP048050.1_1     0  NZ_CP048048.1    44206    42   \n",
       "...                     ...   ...            ...      ...   ...   \n",
       "445330  NZ_CP048049.1_99997     0    NC_009480.1   243049     0   \n",
       "445331  NZ_CP048050.1_99997     0    NC_009480.1  2942610    42   \n",
       "445332    NC_009480.1_99997     0    NC_009479.1    69488    42   \n",
       "445333    NC_009478.1_99997     0    NC_009480.1  1808419     0   \n",
       "445334    NC_009479.1_99997     0    NC_009480.1  2741813    24   \n",
       "\n",
       "                CIGAR       Organims  \n",
       "0                150M   capsici_1106  \n",
       "1                150M   capsici_1106  \n",
       "2                150M   capsici_1106  \n",
       "3                150M   capsici_1106  \n",
       "4                150M   capsici_1106  \n",
       "...               ...            ...  \n",
       "445330  42M1I3M1D104M  michi_contigs  \n",
       "445331           150M  michi_contigs  \n",
       "445332           150M  michi_contigs  \n",
       "445333           150M  michi_contigs  \n",
       "445334           150M  michi_contigs  \n",
       "\n",
       "[1381790 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ruta_tablas = '/home/betterlab/GIT/clavibacter/scripts-tda/output-tda/prueba-3genomas/tablas/'\n",
    "contenido = os.listdir(ruta_tablas)\n",
    "genomes = []\n",
    "df=[]\n",
    "for i in contenido:\n",
    "    dftt = pd.read_csv(ruta_tablas + i, sep='\\t', header=None) #importar las tablas para cada genoma\n",
    "    dftt['Organims']=i[16:-11] #agregar el nombre del organismo\n",
    "    df.append(dftt) #agregar al dataframe\n",
    "    genomes.append(i[16:-11])\n",
    "    \n",
    "df=pd.concat(df)\n",
    "df.columns= ['Qname','Flag', 'Rname', 'Pos','MapQ','CIGAR','Organims'] #cambiar nombre a las columnas del dataframe\n",
    "df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('MapeoReads-organismos.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "En 'df2' agrupamos por nombre del read y por organimos.\n",
    "En 'df3'agrupamos por read y obtenemos a cuantos organismos mapea\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Qname                Organims     \n",
       "NC_009478.1_0        capsici_1101     1\n",
       "                     capsici_1106     1\n",
       "                     michi_contigs    1\n",
       "NC_009478.1_1        michi_contigs    1\n",
       "NC_009478.1_10       capsici_1101     1\n",
       "                                     ..\n",
       "NZ_CP048050.1_99998  capsici_1101     1\n",
       "                     capsici_1106     1\n",
       "NZ_CP048050.1_99999  capsici_1101     1\n",
       "                     capsici_1106     1\n",
       "                     michi_contigs    1\n",
       "Name: Qname, Length: 1381790, dtype: int64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2=df.groupby(['Qname','Organims'])['Qname'].size()\n",
    "df2\n",
    "df3=df.groupby(['Qname'])['Qname'].count().reset_index(name='count')\n",
    "df2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "cannot insert Qname, already exists",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-75-f551e3e264f2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mreset_index\u001b[0;34m(self, level, drop, name, inplace)\u001b[0m\n\u001b[1;32m   1584\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1585\u001b[0m             \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_frame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1586\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdrop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1587\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1588\u001b[0m     \u001b[0;31m# ----------------------------------------------------------------------\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mreset_index\u001b[0;34m(self, level, drop, inplace, col_level, col_fill)\u001b[0m\n\u001b[1;32m   4692\u001b[0m                 \u001b[0;31m# to ndarray and maybe infer different dtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4693\u001b[0m                 \u001b[0mlevel_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_maybe_casted_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlev\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlab\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4694\u001b[0;31m                 \u001b[0mnew_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel_values\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4695\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4696\u001b[0m         \u001b[0mnew_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36minsert\u001b[0;34m(self, loc, column, value, allow_duplicates)\u001b[0m\n\u001b[1;32m   3574\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ensure_valid_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3575\u001b[0m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbroadcast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3576\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_duplicates\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_duplicates\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3577\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3578\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0massign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36minsert\u001b[0;34m(self, loc, item, value, allow_duplicates)\u001b[0m\n\u001b[1;32m   1171\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mallow_duplicates\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1172\u001b[0m             \u001b[0;31m# Should this be a different kind of error??\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1173\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"cannot insert {}, already exists\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1174\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1175\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: cannot insert Qname, already exists"
     ]
    }
   ],
   "source": [
    "df2.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "X=df['Organims'].unique().tolist()\n",
    "Xset(X)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Definimos la funcion potencia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def powerset(fullset):\n",
    "  listsub = list(fullset)\n",
    "  subsets = []\n",
    "  for i in range(2**len(listsub)):\n",
    "    subset = []\n",
    "    for k in range(len(listsub)):            \n",
    "      if i & 1<<k:\n",
    "        subset.append(listsub[k])\n",
    "    subsets.append(subset)        \n",
    "  return subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[],\n",
       " ['michi_contigs'],\n",
       " ['capsici_1101'],\n",
       " ['michi_contigs', 'capsici_1101'],\n",
       " ['capsici_1106'],\n",
       " ['michi_contigs', 'capsici_1106'],\n",
       " ['capsici_1101', 'capsici_1106'],\n",
       " ['michi_contigs', 'capsici_1101', 'capsici_1106']]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "powerset(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Qname</th>\n",
       "      <th>Flag</th>\n",
       "      <th>Rname</th>\n",
       "      <th>Pos</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>Organims</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NZ_CP048050.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>45254</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC_009480.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009478.1</td>\n",
       "      <td>10211</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC_009478.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2947543</td>\n",
       "      <td>23</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NZ_CP048047.1_0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>13579</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NZ_CP048049.1_1</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>29946</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445330</td>\n",
       "      <td>NZ_CP048049.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>243049</td>\n",
       "      <td>0</td>\n",
       "      <td>42M1I3M1D104M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445331</td>\n",
       "      <td>NZ_CP048050.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2942610</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445332</td>\n",
       "      <td>NC_009480.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009479.1</td>\n",
       "      <td>69488</td>\n",
       "      <td>42</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445333</td>\n",
       "      <td>NC_009478.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>1808419</td>\n",
       "      <td>0</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>445334</td>\n",
       "      <td>NC_009479.1_99997</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_009480.1</td>\n",
       "      <td>2741813</td>\n",
       "      <td>24</td>\n",
       "      <td>150M</td>\n",
       "      <td>michi_contigs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>445335 rows √ó 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Qname  Flag        Rname      Pos  MapQ          CIGAR  \\\n",
       "0           NZ_CP048050.1_0     0  NC_009479.1    45254    42           150M   \n",
       "1             NC_009480.1_0     0  NC_009478.1    10211    42           150M   \n",
       "2             NC_009478.1_0     0  NC_009480.1  2947543    23           150M   \n",
       "3           NZ_CP048047.1_0     0  NC_009479.1    13579    42           150M   \n",
       "4           NZ_CP048049.1_1     0  NC_009480.1    29946    42           150M   \n",
       "...                     ...   ...          ...      ...   ...            ...   \n",
       "445330  NZ_CP048049.1_99997     0  NC_009480.1   243049     0  42M1I3M1D104M   \n",
       "445331  NZ_CP048050.1_99997     0  NC_009480.1  2942610    42           150M   \n",
       "445332    NC_009480.1_99997     0  NC_009479.1    69488    42           150M   \n",
       "445333    NC_009478.1_99997     0  NC_009480.1  1808419     0           150M   \n",
       "445334    NC_009479.1_99997     0  NC_009480.1  2741813    24           150M   \n",
       "\n",
       "             Organims  \n",
       "0       michi_contigs  \n",
       "1       michi_contigs  \n",
       "2       michi_contigs  \n",
       "3       michi_contigs  \n",
       "4       michi_contigs  \n",
       "...               ...  \n",
       "445330  michi_contigs  \n",
       "445331  michi_contigs  \n",
       "445332  michi_contigs  \n",
       "445333  michi_contigs  \n",
       "445334  michi_contigs  \n",
       "\n",
       "[445335 rows x 7 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.loc[:,'Organims']=='michi_contigs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'Qname'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-b20d84ddf9c5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mQname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5177\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5178\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5179\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5181\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'Qname'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
